### DISCUSSION TITLE
Feedback on my similar tool

### DISCUSSION THREAD

**Discussion by covert-encryption** [samuel-lucas6/Kryptor#32] [General]
You should be aware of our effort at finally getting encryption right (and killing PGP). First and foremost, our files have no identifiable header bytes: without correct keys it looks like random data. Yes, it was a little hard to design, but we got it done.

GPG, Age, Kryptor and yes, everybody else reveal the exact size of the plaintext because the encrypted file only adds its headers. While random padding is not a new idea, we implement it correctly while others don't support it at all. We can also put a message and a bunch of files into a single archive (not possible with most other tools, Krypton included as I understand), which hides the sizes and count of individual files, and allows for easier transport of the whole package.

Even with the padding included, our messages are super short, only one line of Base64 where GPG and Age use about six lines, and the gap only widens once you switch to public keys or perhaps have many recipients and signatures.

Within weeks, I claim, we have implemented more features and for a large part better than any of our competition. The console UI is already pretty good. The encryption runs faster than any other program that we know of (around 3 GB/s).

Covert Encryption is still in early development, but we want people to test it and give feedback, especially on the specification because all changes are still possible. For Windows we plan to do a GUI soon and thus won't bother to fix operation in legacy terminals, so use Windows Terminal or other modern options if you wish to try it at this point. Other platforms have no such issues.

https://github.com/covert-encryption/covert

**Comment by samuel-lucas6** [comment]
This is a very advertisey comment, and from a quick look, I can already see some questionable things and comments going on. For example, password hashing for dozens of seconds is completely impractical, forward secrecy isn't possible using a non-interactive key exchange, the same nonce is being used for various things, you're XORing keys, and you've said that using numbers and symbols weakens a passphrase. I'm afraid I don't currently have the time to look into your tool properly.

I'm going to investigate implementing an indistinguishable from random format in v4 because minimising metadata is a goal of mine. However, it's a lot harder to do. One of the main issues would be applying Elligator to ephemeral public keys and padding messages in a way to limit leakage.

As for the current version of Kryptor, it does not reveal the exact size of the plaintext because it encrypts in 16 KiB chunks. Even if you encrypt a 0 byte file, it will be padded to 16 KiB. Age also performs chunking but only for larger file sizes. Directory encryption does not currently result in one file, but that's another thing I'm considering for v4. It does have various disadvantages though, like not being able to decrypt or back up individual files in the directory. A corruption could prevent decryption of the entire directory, whereas encrypting files individually means a corruption is less of a problem. It's also again more difficult to implement.

**Reply by covert-encryption** [reply]
Thanks for your comments and for checking out the specification, and in particular the security document where we review the potential issues that you mention, like that the ephemeral public key needs to be scrambled (implementing elligator avoids a 2 bit information leak per file of it potentially being encrypted in this scheme).

Forward secrecy actually is possible, although not as practical as with an Internet protocol. A keyserver or whichever system you use to acquire the other party's public key before sending the first message can use [X3DH](https://signal.org/docs/specifications/x3dh/), and even without that any reply messages can use a temporary key (ratchet) established in the first message. It only needs some local storage for these keys, no online connection.

I would suggest also looking into random padding in v4. Granted, padding to fixed blocks gives some security over content length, but for a set of files you still mostly reveal which set of files it was with high confidence, and with such a large size you make it impractical to use on online forums in ASCII armored format - not sure if that is your intended use case but Covert attempts to cover anything from messages of a few bytes to files of tens of gigabytes, and apparently we can without any compromise. 

Without random padding, adding just a single byte may make the ciphertext longer, which can be enough when the attacker knows the possible plain text lengths. Also, if all your files are a multiple of a known block length (perhaps plus a known header size), the lengths of the files can still reveal the format despite obscuring the header bytes. The only way to solve these two issues is to use randomness too (we use padding to a fixed minimum size always combined with random).

Regarding password security, 8 characters is really not enough. Let's say you want to be secure and require uppercase, digits and symbols, but still let people choose their own password, you are very likely to get the pattern Xxxxxx1# where x represent some letter of the alphabet (likely forming a word or a slight variation of), 1 represents one of the ten digits and # represents some of the handful of common symbols on the keyboard, meaning that this password only offers about 18 bits of security and it can be easily brute forced if hashing does not take a very long time. Even if a password generator meeting these requirements was used, the requirement of numbers and symbols means that at least two of the characters are constrained to about 10 choices rather than the 52 offered by upper and lower case alphabet alone. The reasoning behind this requirement is that the alphabet should be larger, so that you get 72^8 options for the 8-character password but because of the restriction itself you actually get far fewer (i.e. requiring digits and symbols actually is BAD). The comic XKCD popularised this problem long ago, and even [NIST guidelines](https://pages.nist.gov/800-63-3/sp800-63b.html#appA) are finally [changing](https://www.excaltech.com/understanding-the-new-nist-guidelines-for-password-security/) this stupid requirement everyone held for decades.

![](https://imgs.xkcd.com/comics/password_strength.png)
[XKCD Password Strength](https://xkcd.com/936/)

We require a minimum of 8 characters and for that length use very strong hashing so that it cannot be broken within hours by anyone willing to pay a little to use AWS cloud to do a brute force attack. You can also easily avoid the delay simply by using a password of 9 or 10 letters, or of four random words as we recommend. Or by using public keys, to avoid typing and delays entirely, if you can keep the secret key safe (great for sensitive log files and other things where automated systems need to encrypt something holding only the public key).

**Reply by samuel-lucas6** [reply]
I'm not sure what you mean by 'of it potentially being encrypted in this scheme'.

Yes, it's possible when there are preshared one-time keys, but in the context of a program like Kryptor, that's not really doable. There are also the issues of erasing keys properly and having to store multiple keys. Technically, you can already obtain forward secrecy by just deleting the keys used each time.

I'll probably look into something like [PADME](https://bford.info/pub/sec/purb.pdf) padding more. An ASCII armour format was never a design goal because I've rarely seen such formats used. They're more of a historical thing since we can now use files in most cases. I'm also trying not to offer too many commands.

I agree that 8 character passwords should not be used. Therefore, surely a better solution is just to limit the minimum length to something like 12 characters? Using numbers and symbols should certainly not be discouraged for longer passwords.

**Reply by covert-encryption** [reply]
Given a file with an exposed key that you can verify as a valid point, there is a 25 % chance of that being random data rather than a key. Given many files, this becomes statistically significant, and thus needs to be prevented.

We actually have the PADME white paper linked on our front page, but that scheme is flawed in that it fails to add randomness, and it also wastes quite a lot of space while producing files of only specific byte lengths that can then be identified.

**Reply by samuel-lucas6** [reply]
Thank you for the clarification. I've only heard good things about PADME, so I'll look into that.

**Reply by covert-encryption** [reply]
Thanks for the motivation, I just implemented Elligator2. Meanwhile someone made a GUI which is looking pretty cool, hoping to see that PR finished and merged soon. I was expecting it to be until early next year that work on GUI should be started but I definitely don't mind having one now.

Regarding padding, since everyone seems so hyped about Padme, I did have a closer look. It turns out that the overhead is actually very low, and that the 12 % figure was only the maximal peak value. As for other features, here are a few diagrams to illustrate.

![Padding size](https://github.com/covert-encryption/covert/raw/main/docs/in-out.png)
Data shown in gray, padding on top of it in orange, and in case of Covert the fading shades of orange represent randomness. For each output size, we should have many possible input sizes that could have produced it.

![Output size distribution](https://github.com/covert-encryption/covert/raw/main/docs/distribution.png)
If only clearly distinct known output sizes are produced, an adversary can find out which padding scheme was used. We do not want those spikes. Even a single larger file can be revealing. E.g. 1114112 bytes is almost certainly padme padding because it is one of the 33 sizes that padme uses within the 1-2 MiB range, where you would expect a million different sizes if they were random.

**Reply by samuel-lucas6** [reply]
Yes, the minimal overhead seems to be the main selling point. Nice graphs thanks. I do agree that deterministic padding is more problematic. 

When padding with chunked encryption, you can either:
1. Pad each input chunk.
2. Pad each output chunk.
3. Pad based on the entire input size.
4. Pad based on the entire output size.

The first two require storing more information about the padding and complicate decryption. The bottom two result in having to store less padding information, and the padding can be easily skipped or removed. Which would you recommend? I presume you're calculating the amount of padding based on the input file size rather than the output size.

**Reply by covert-encryption** [reply]
Covert indices all the input files and reads their names and sizes (if available) before starting, writing an index at the beginning. If all sizes are known, the padding size can be calculated before starting encryption, but if there are any streaming inputs (e.g. from a pipe), we do that after all the files have been written so that the total content size is known. It is also possible to add padding in the middle of streaming, which may be necessary to fill a block when less data is available than the current block size, and real time operation is needed.

The padding size is calculated on *content only* to allow short messages to be protected while keeping the output size under 100 bytes. An adversary who knows that a short file is a covert message can discover that it can only have one recipient and no signatures (as having any more data would make it larger than the actual size found), just as she can deduce the maximum size of content that there may possibly be. However, randomness makes it far less likely that anything of importance could be extracted unless the adversary can have the same message encrypted many times, allowing statistical analysis, or if she gets lucky and gets a random size that matches only one of the known possible plain text lengths. Any deterministic padding that still allows the output size adapt to content still reveals more. I also attempted combining Padme with randomness either before or after it but either way it was worse than our current system.

Padding is best added at the end, after all the content. To keep every byte authenticated and to avoid having to generate random bytes we put the padding inside ChaCha20-Poly1305 and always read all of it (another implementation can choose to stop early of course, just as they can seek to skip to a specific file while decrypting). If you don't care about full file authentication, you can simply add random bytes after your format has already indicated that the data has ended.

Full authentication is relevant when an adversary can try altering some of the bytes and then test with a decoding oracle whether that causes an error. If your padding is not authenticated, any modification to those bytes goes undetected, allowing Malice to learn the amount of padding there is in that file. A similar problem with the auth header bytes, where altering any bytes actually used just causes a decryption failure, so you might think you need no authentication. However, if altering some of the bytes causes no error, Malice learns that those bytes are part of the auth header and that they contain a recipient other than the one she is testing on. Thus, you really should have every byte authenticated, which of course is very easy to do with AEAD ciphers, no reason to use HMACs anymore.

**Reply by samuel-lucas6** [reply]
Pad-then-Encrypt seems sensible. I agree that every byte of the file should be authenticated. Prepending or appending would work for Encrypt-then-Pad though if you authenticated those bytes by using them as additional data. That's what I was getting at. The official [PURBs](https://github.com/dedis/purb/pull/11) implementation switched from Pad-then-Encrypt to Encrypt-then-Pad.

I'm afraid I don't agree with you about HMAC though. Encrypt-then-MAC provides better security than an AEAD when implemented correctly. The main disadvantage is of course speed, especially compared to newer algorithms like [AEGIS](https://github.com/jedisct1/rust-aegis) and [ROCCA](https://github.com/jedisct1/rust-rocca).

**Reply by hakavlad** [reply]
Could you please take a look at this new tool?
https://github.com/hakavlad/tird

It's a file encryption tool focused on minimizing metadata and hiding encrypted data.
A draft specification is available. The code is documented.

**Reply by samuel-lucas6** [reply]
> Could you please take a look at this new tool?
> https://github.com/hakavlad/tird

It certainly looks very interesting. Props for documenting it. I'll try to read through the documentation properly at a later date.
